### PR TITLE
Experiment/scu

### PR DESCRIPTION
**Pull request by boygirl**
cc/ @chrisbolin 

This was the most fool-proof way I could figure out to prevent unnecessary re-renders... 

sCU logic is roughly:
animating ? true
state changes* ? true
props changes* ? true
else false

state changes:
calculates all the aggregate state changes that could affect a given component in sCU and compares them to the existing cached state changes in a deep equality check. State changes on the `parent` namespace are ignored when the component is not standalone.

props changes:
prior to a deep equality check on props, and functions are removed. We were seeing unnecessary re-renders because isEqual doesn't support checking functions, and always returns false when comparing functions. 

This change requires adding `options` to non-standard Victory components (i.e. axes and continuous components like `VictoryArea`) so that the state check know which types / indices to check i.e. "data", "labels", "tickLabels" etc.
See: https://github.com/FormidableLabs/victory-chart/pull/500 (whoa, 500!)

This change is still slower than your WIP branch, but hits all the edge cases that were missing, specifically updating parent states (i.e. stacked bar example in victory voronoi container demo). I'm seeing what can be done to speed it up.

### CODE REVIEW

**Comment by boygirl** [comment]
Update: 
I changed from checking on all calculated props, to performing a deep quality check strictly on props, and another deep equality check on aggregated state changes. Performance is much better. I think the slow step at this part is the deep equality check on the data prop itself, but I'm not really sure of a way around that one.

**Comment by chrisbolin** [comment]
awesome. from what I can see it is now functionally correct (unlike my attempt).

but you're right, the heavily lifting is still quite heavy. `this.getCalculatedValues(nextProps)` seems to be the biggest burden in my testing. I am guessing it does some "work" that scales with `data` size.

Are we copying `props.data`? If so, not only would the copy take a long time, but the equality check will take _much_ longer. This is because `isEqual` can "cheat" when comparing two large arrays by first seeing if they are simply references to the same array in memory. If that doesn't work, it will have to iterate through every element. Usually this 2nd process doesn't take too long, because it probably uses some good heuristics (like check last element, check length, etc). But if you copy an identical array and then compare it to its source, it will take a long time to prove its equality.

**Comment by chrisbolin** [comment]
`cacheValues` is very clever -- reuse the values calculated in sCU in the render. awesome :)

**Comment by boygirl** [comment]
@chrisbolin it's still a noticeable improvement over master for large datasets. I'm not sure how to get around the slow step.

File: src/victory-util/add-events.js [hunk 1]
```diff
@@ -1,9 +1,41 @@
...
 import Events from "./events";
 import VictoryTransition from "../victory-transition/victory-transition";
 
-export default (WrappedComponent) => {
+//  used for checking state changes. Expected components can be passed in via options
+const defaultComponents = [
+  { name: "parent", index: "parent" }, { name: "data" }, { name: "labels" }
+];
+
+const areVictoryPropsEqual = (a, b) => {
```
Comment (ryan-roemer) [inline src/victory-util/add-events.js:13]: Something like this should have a ton of unit tests behind it. A recursive equality check is something that it's so easy to have incorrect edge cases with.

Perhaps ticket out for "soon" if not now?

/cc @boygirl @chrisbolin

File: src/victory-util/add-events.js [hunk 2]
```diff
@@ -14,30 +46,86 @@ export default (WrappedComponent) => {
...
+
+      options = options || {};
+      const components = options.components || defaultComponents;
+      return components.map((component) => {
+        if (!props.standalone && component.name === "parent") {
+           // don't check for changes on parent props for non-standalone components
+          return undefined;
+        } else {
+          return component.index !== undefined ?
+          getState(component.index, component.name) :
```
Comment (ryan-roemer) [inline src/victory-util/add-events.js:99]: NIT: indentation

File: src/victory-util/add-events.js [hunk 3]
```diff
@@ -14,30 +46,86 @@ export default (WrappedComponent) => {
...
+      };
+
+      options = options || {};
+      const components = options.components || defaultComponents;
+      return components.map((component) => {
+        if (!props.standalone && component.name === "parent") {
+           // don't check for changes on parent props for non-standalone components
+          return undefined;
+        } else {
+          return component.index !== undefined ?
```
Comment (ryan-roemer) [inline src/victory-util/add-events.js:98]: NIT: `typeof FOO === "undefined"` or `_.isUndefined()`

File: src/victory-util/add-events.js [hunk 4]
```diff
@@ -14,30 +46,86 @@ export default (WrappedComponent) => {
...
-      this.getSharedEventState = sharedEvents && isFunction(sharedEvents.getEventState) ?
+      const componentEvents = Events.getComponentEvents(props, components);
+      const getSharedEventState = sharedEvents && isFunction(sharedEvents.getEventState) ?
         sharedEvents.getEventState : () => undefined;
-      this.baseProps = this.getBaseProps(props);
-      this.dataKeys = Object.keys(this.baseProps).filter((key) => key !== "parent");
-      this.hasEvents = props.events || props.sharedEvents || this.componentEvents;
-      this.events = this.getAllEvents(props);
+      const baseProps = this.getBaseProps(props, getSharedEventState);
+      const dataKeys = Object.keys(baseProps).filter((key) => key !== "parent");
```
Comment (ryan-roemer) [inline src/victory-util/add-events.js:112]: You're importing lodash `_.keys` elsewhere. Perhaps decide on one canonical way to do it? (Unless there's some reason for `_.keys` above that `Object.keys` can't do).

File: src/victory-util/add-events.js [hunk 5]
```diff
@@ -14,30 +46,86 @@ export default (WrappedComponent) => {
...
+      return {
+        componentEvents, getSharedEventState, baseProps, dataKeys, hasEvents, events
+      };
     }
 
-    getBaseProps(props) {
-      const sharedParentState = this.getSharedEventState("parent", "parent");
+    cacheValues(obj) {
+      keys(obj).forEach((key) => {
+        this[key] = obj[key];
```
Comment (ryan-roemer) [inline src/victory-util/add-events.js:122]: Are there any collision protections? It seems like you're adding straight to the underlying object. E.g., if someone  passes an object with:

```js
{ cacheValues: "foo" }
```

it would blow away this version method of `this.cacheValues()`

File: src/victory-util/events.js [hunk 6]
```diff
@@ -125,7 +125,7 @@ export default {
         const mutationKeys = getKeys(childName);
         return Array.isArray(mutationKeys) ?
           mutationKeys.reduce((memo, key) => {
-            return assign(memo, getMutationObject(key, childName));
+            return assign({}, memo, getMutationObject(key, childName));
```
Comment (ryan-roemer) [inline src/victory-util/events.js:128]: Why aren't we mutating `memo`? Seems like extra work to copy all the props for no benefit (?)

File: src/victory-util/events.js [hunk 7]
```diff
@@ -134,7 +134,7 @@ export default {
       const allChildNames = childNames === "all" ?
         without(Object.keys(baseProps), "parent") : childNames;
       return Array.isArray(allChildNames) ? allChildNames.reduce((memo, childName) => {
-        return assign(memo, getReturnByChild(childName));
+        return assign({}, memo, getReturnByChild(childName));
```
Comment (ryan-roemer) [inline src/victory-util/events.js:137]: Same here.
